### DISCUSSION TITLE
I have clangd installed. How do I connect it with ALE?

### DISCUSSION THREAD

**Discussion by ubaldot** [w0rp/ale#3937] [Q&A - Ask for help with problems] [answered]
When I use commands like `:ALEGoToDefinition`, `:ALEComplete`, etc, nothing seems to happen.
I am suspicious that `ALE` is simply not connected to `clangd` (in-spite `clang-tidy` works amazingly).
Inspection of `:help ale-lint-language-server` suggest to add the following to my `_vimrc`. 

```
call ale#linter#Define('c', {
  \   'name': 'clangd',
  \   'lsp': 'stdio',
  \   'executable': 'C:\Program Files (x86)\LLVM\bin',
  \   'command': '%e run',
  \   'project_root':'C:\MyProject',
  \})
```

But when I run `:ALEGoToDefinition` of `:ALEComplete` nothing happens. 
I have a `compile_commands.json` file and `<C-x> <C-o> `works well, but `ALE`. 

My `Vim`:

```
VIM - Vi IMproved 8.2 (2019 Dec 12, compiled Sep 21 2021 22:13:38)
MS-Windows 32-bit console version
Included patches: 1-3452
Compiled by appveyor@APPVYR-WIN
```

My `ALEInfo`

```
 Current Filetype: c
Available Linters: ['cc', 'ccls', 'clangd', 'clangtidy', 'cppcheck', 'cquery', 'flawfinder']
   Linter Aliases:
'cc' -> ['gcc', 'clang']
  Enabled Linters: ['clangtidy']
  Ignored Linters: []
 Suggested Fixers: 
  'astyle' - Fix C/C++ with astyle.
  'clang-format' - Fix C/C++ and cuda files with clang-format.
  'clangtidy' - Fix C/C++ and ObjectiveC files with clang-tidy.
  'remove_trailing_lines' - Remove all blank lines at the end of a file.
  'trim_whitespace' - Remove all trailing whitespace characters at the end of every line.
  'uncrustify' - Fix C, C++, C#, ObjectiveC, ObjectiveC++, D, Java, Pawn, and VALA files with uncrustify.
 Linter Variables:

let g:ale_c_always_make = 0
let g:ale_c_build_dir = ''
let g:ale_c_build_dir_names = ['build', 'bin']
let g:ale_c_clangtidy_checks = ['*']
let g:ale_c_clangtidy_executable = 'clang-tidy'
let g:ale_c_clangtidy_extra_options = '--extra-arg=-I --extra-arg=C:\Devtools\Install\Cygwin\3.0.7\usr\include --extra-arg=-I --extra-arg=C:\PE_Builds\sw-ecu\BSW\_out\P_AMTG_MC_TEA2P\bin\all_inc'
let g:ale_c_clangtidy_options = ''
let g:ale_c_parse_compile_commands = 1
let g:ale_c_parse_makefile = 0
 Global Variables:

let g:ale_cache_executable_check_failures = v:null
let g:ale_change_sign_column_color = 0
let g:ale_command_wrapper = ''
let g:ale_completion_delay = 100
let g:ale_completion_enabled = 0
let g:ale_completion_max_suggestions = 50
let g:ale_disable_lsp = 0
let g:ale_echo_cursor = 1
let g:ale_echo_msg_error_str = 'Error'
let g:ale_echo_msg_format = '%code: %%s'
let g:ale_echo_msg_info_str = 'Info'
let g:ale_echo_msg_warning_str = 'Warning'
let g:ale_enabled = 1
let g:ale_fix_on_save = 0
let g:ale_fixers = {'c': ['astyle', 'clangtidy', 'remove_trailing_lines', 'trim_whitespace']}
let g:ale_history_enabled = 1
let g:ale_history_log_output = 1
let g:ale_keep_list_window_open = 0
let g:ale_lint_delay = 200
let g:ale_lint_on_enter = 1
let g:ale_lint_on_filetype_changed = 1
let g:ale_lint_on_insert_leave = 1
let g:ale_lint_on_save = 1
let g:ale_lint_on_text_changed = 'normal'
let g:ale_linter_aliases = {}
let g:ale_linters = {'c': ['clangtidy'], 'python': ['flake8']}
let g:ale_linters_explicit = 0
let g:ale_linters_ignore = {}
let g:ale_list_vertical = 0
let g:ale_list_window_size = 10
let g:ale_loclist_msg_format = '%code: %%s'
let g:ale_max_buffer_history_size = 20
let g:ale_max_signs = -1
let g:ale_maximum_file_size = v:null
let g:ale_open_list = 0
let g:ale_pattern_options = v:null
let g:ale_pattern_options_enabled = v:null
let g:ale_root = {}
let g:ale_set_balloons = 0
let g:ale_set_highlights = 1
let g:ale_set_loclist = 1
let g:ale_set_quickfix = 0
let g:ale_set_signs = 1
let g:ale_sign_column_always = 0
let g:ale_sign_error = '>>'
let g:ale_sign_info = '--'
let g:ale_sign_offset = 1000000
let g:ale_sign_style_error = '>>'
let g:ale_sign_style_warning = '--'
let g:ale_sign_warning = '--'
let g:ale_sign_highlight_linenrs = 0
let g:ale_statusline_format = v:null
let g:ale_type_map = {}
let g:ale_use_global_executables = v:null
let g:ale_virtualtext_cursor = 0
let g:ale_warn_about_trailing_blank_lines = 1
let g:ale_warn_about_trailing_whitespace = 1
  Command History:

(executable check - success) clang-tidy
(finished - exit code -1073741795) 'cmd /s/c "clang-tidy -checks=* "--extra-arg=-I --extra-arg=C:\Cygwin\3.0.7\usr\include --extra-arg=-I --extra-arg=C:\MyProject\_out\all_inc" C:\MyProject\src\MySource.c -p C:\MyProject"'

<<<NO OUTPUT RETURNED>>>

(started) 'cmd /s/c "clang-tidy -checks=* "--extra-arg=-I --extra-arg=C:\Cygwin\3.0.7\usr\include --extra-arg=-I --extra-arg=C:\MyProject\_out\all_inc" C:\MyProject\src\MySource.c  -p C:\MyProject"'

```

**Comment by hsanson** [comment, marked as answer]
The clangd is already defined in ALE, you do not need to define it again in your configuration. 

First add `clangd` to your list of enabled linters. Currently you only have defined clangtidy for C files (see `:h g:ale_linters`):

```
let g:ale_linters = {'c': ['clangtidy'], 'python': ['flake8']}
```

ALE should be able to find the clangd binary on its own but you can tell it where to find it if it doesn't:

```
let g:ale_c_clangd_executable = '/path/to/bin/clangd'
```

You can also set options that ALE will use when invoking clangd. See `:h ale-c-clangd` for additonal details.

Finally make sure vim/neovim current folder is a C project folder. ALE won't start the language server if it cannot find the project root. The project root will be the path where the compile_commands.json file is located.

**Reply by ubaldot** [reply]
Perfect!
Now `clangd` is recognized well. Many thanks!

However,  `:ALEGoToDefinition` and `:ALEFindReferences ` seems to partially work.
More precisely, they can find definitions and references only among opened buffers. 
For example, if ` MyFile1.c` calls `foo()` and `foo()` is defined in `MyFile2.c` (contained in the same folder), then  `:ALEGoToDefinition` will not find `foo() ` if `MyFile2.c` is not opened. 

Also, `:ALEComplete` does not seem to work super good.
If I have a user-defined function (say `cdd_foo()`) and then I type `cdd_ `and then I call `:ALEComplete` nothing happens if `cdd_foo() `is declared in a file not located in the same folder.  
Otherwise, the completion seems to work for all those functions whose declaration is "in-folder-scope" (for example in some `.h` file located in the same folder of my currently opened `.c` file).
If it normal? How to fix it? 
Or maybe this is the due to the double quotation mark issue that I have?
More precisely, from my `:ALEInfo `of above you can see that  `clang-tidy` delimits the various  `--extra-arg` with a double quotation mark, `"` i.e.  

`(finished - exit code -1073741795) 'cmd /s/c "clang-tidy -checks=* "--extra-arg=-I --extra-arg=C:\Cygwin\3.0.7\usr\include --extra-arg=-I --extra-arg=C:\MyProject\_out\all_inc" C:\MyProject\src\MySource.c -p C:\MyProject"' 
`

This leads to several linting errors (and perhaps to the fact that I cannot find functions defined in some other file, see above). 
I tried to run the same command from the `Powershell` without any quotation marks, i.e. 

`clang-tidy -checks=* --extra-arg=-I --extra-arg=C:\Cygwin\3.0.7\usr\include --extra-arg=-I --extra-arg=C:\MyProject\_out\all_inc C:\MyProject\src\MySource.c -p C:\MyProject`

and it works correctly.
I have raised a bug for it but it seems to have been forgotten :/

**Reply by hsanson** [reply]
Very difficult for me to help you here.  I do not use clangd, last time I did C was 10 years ago, and I do not have any Windows machines around me.

Would recommend you create a small "hello world" C project that presents the same problems and share it in a git repo. Maybe someone with similar environment as you can help debug what may be the problem.

**Reply by ubaldot** [reply]
No problem at all! :)
You already hugely helped me!
Many thanks!

Regarding the jump to definition/declaration/implementation/... I have found this:

https://github.com/dense-analysis/ale/issues/2550

Regarding the double quotation mark added in the `--extra-arg`  I will wait. 
In the meantime I will run `clang-tidy` from` Powershell` to figure what is going on with my code :)
